contents/java/Java.md

```diff
@@ -14,6 +14,161 @@ Authors: John Doe, Jane Doe
+
+```java
+Class<Student> studentClass = Student.class;
+```
+
+It also means that you can store any Class object in any data structure for future retrieval. This is the main entry
+point for Java's reflections. You can now get the name of the class, create new instances of the class, observe its
+public/private fields - the possibilities are endless!
+
+### Getting Started
+
+There are many webpages dedicated to explaining the details of reflections in Java; so this will not repeat what is
+being made readily available on the web. One good place to start is [http://www.javaworld.com/article/2077015/java-se/take-an-in-depth-look-at-the-java-reflection-api.html](http://www.javaworld.com/article/2077015/java-se/take-an-in-depth-look-at-the-java-reflection-api.html)
+
+What this page will do however is to give some examples to demonstrate what reflections can do and in then give 
+you ideas on how you can use reflections in your project.
+
+One important point is that while Java reflections are powerful, its implementations are anything but. There are however
+some libraries out there to make life easier such as the [Google's Guava
+library](https://github.com/google/guava/wiki/ReflectionExplained) which contains many utility methods and classes to
+make your life easier.
+
+#### Applications
+
+You might have learnt from your Software Engineering module that the Observer pattern can be used for Objects that are
+interested to get notified if the state of another object is changed. However, a you don't want to create a troublesome
+bidirectional dependency between two unrelated objects that have no business talking to each other. This is when the
+Observer pattern becomes useful.
+
+One prime example of the implementation of the `Observer` pattern is the Google Events bus used in `AddressBook Level 4`.
+The event bus uses reflections to observe all registered objects via `register` method for methods annotated with the
+`Subscribe` annotation.
+
+An example implementation (not the actual) of the `Subscribe` annotation might be:
+
+```java
+@Retention(RetentionPolicy.RUNTIME) // Retain this annotation at runtime!
+@Target(ElementType.METHOD) // Only can be applied to methods!
+public @interface Subscribe { }
+```
+
+And that is it! Important parts of the code to note is the first and second line before the declaration. The first line
+tells Java that this annotation must not be discarded during compile time so it will be available during runtime. The
+retention poilcy is there because some annotation do not mean anything after compilation (such as `Override` and
+`SuppressWarnings`), so it does not make sense to keep it after compiling.
+
+The second line just means that this annotation can be applied to methods only.
+
+And the more important part is the Subscriber Registry
+
+```java
+// TypeToken class is provided by Google Guava reflection library
+Set<? extends Class<?>> supertypes = TypeToken.of(clazz).getTypes().rawTypes();
```

Comment:
Maybe you can instantiate a `Sheep` object here and give the object a more meaningful name here!